File: src/main/java/org/BeehiveRobotics/Library/Motors/Motor.java
package org.BeehiveRobotics.Library.Motors;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.util.Range;
import java.util.Runnable;

public class Motor implements Runnable {
    private static final double RAMP_LOG_EXPO = 0.8;
    private double MIN_SPEED = 0.2;
    private double MAX_SPEED = 1;
    private DcMotor motor;
    private MotorModel model;
    private String name;
    private LinearOpMode opMode;
    private double target;
    private double current;
    private double power;

    Motor(LinearOpMode opMode, String name) {
        this.name = name;
        this.opMode = opMode;
        this.motor = this.opMode.hardwareMap.get(DcMotor.class, name);
    }

    Motor setRunMode(DcMotor.RunMode runMode) {
        this.motor.setMode(runMode);
        return this;
    }

    Motor setModel(MotorModel motorModel) {
        this.model = motorModel;
        return this;
    }

    Motor setMinSpeed(double speed) {
        this.MIN_SPEED = Math.abs(speed);
        return this;
    }

    public Motor setMaxSpeed(double speed) {
        this.MAX_SPEED = Math.abs(speed);
        return this;
    }

    Motor resetEncoder() {
        DcMotor.RunMode initialBehavior = this.getRunMode();
        this.setRunMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        this.setRunMode(initialBehavior);
        return this;
    }

    double getCurrentPosition() {
        return this.motor.getCurrentPosition();
    }

    Motor setZeroPowerBehavior(DcMotor.ZeroPowerBehavior zeroPowerBehavior) {
        this.motor.setZeroPowerBehavior(zeroPowerBehavior);
        return this;
    }

    void setTarget(double target) {
        this.target = Math.abs(target);
    }
    
    void runToTarget(double target, double power, boolean waitForStop) {
        this.target = Math.abs(target);
        this.power = power;
        if(!waitForStop) {
            Thread thread = new Thread(this);
            thread.start();
        } else {
            while(!this.isAtTarget()) {
                setPower(power);
            }
        }
    }
    
    void runToTarget(double target, double power) {
        this.runToTarget(target, power, true);
    }

    void setRawPower(double power) {
        if (opMode.opModeIsActive()) {
            if (power > 0) {
                this.motor.setPower(Range.clip(power, MIN_SPEED, MAX_SPEED));
            } else if (power < 0) {
                this.motor.setPower(Range.clip(power, -MAX_SPEED, -MIN_SPEED));
            } else {
                stopMotor();
            }
        } else {
            stopMotor();
        }
    }

    void setPower(double power) {
        this.power = power;
        if (!(this.opMode.opModeIsActive()) || power == 0) {
            stopMotor();
            return;
        }
        this.current = getCurrentPosition();
        double k = 4 / target;
        double calculated_power = k * this.current * (1 - current / target) * power + Double.MIN_VALUE;
        double expo_speed = Math.pow(Math.abs(calculated_power), RAMP_LOG_EXPO);
        if (power < 0) {
            setRawPower(-expo_speed);
            return;
        }
        setRawPower(expo_speed);
    }

    void setPower(double power, double current, double target) {
        this.power = power;
        if (!(this.opMode.opModeIsActive()) || power == 0) {
            stopMotor();
            return;
        }
        this.current = current;
        this.target = target;
        double k = 4 / target;
        double calculated_power = k * this.current * (1 - current / target) * power + Double.MIN_VALUE;
        double expo_speed = Math.pow(Math.abs(calculated_power), RAMP_LOG_EXPO);
        if (power < 0) {
            setRawPower(-expo_speed);
            return;
        }
        setRawPower(expo_speed);
    }

    void stopMotor() {
        this.motor.setPower(0);
    }

    DcMotor.RunMode getRunMode() {
        return this.motor.getMode();
    }

    String getName() {
        return this.name;
    }

    boolean isAtTarget() {
        return Math.abs(current) >= Math.abs(target);
    }

    public Motor setDirection(DcMotorSimple.Direction direction) {
        this.motor.setDirection(direction);
        return this;

    }
    
    public void run() {
        while(!this.isAtTarget() && this.opModeIsActive()) {
            this.setPower(this.power);
        }
        stopMotor();
        Thread.currentThread().interrupt();
    }
}

File: src/main/kotlin/org/BeehiveRobotics/Library/Motors/Motor.kt
package org.BeehiveRobotics.Library.Motors

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.util.Range
import com.qualcomm.robotcore.util.ElapsedTime
import org.BeehiveRobotics.Library.Util.BROpMode
import org.BeehiveRobotics.Library.Systems.RobotSystem

class Motor(private val opMode: BROpMode, val name: String): RobotSystem(opMode), Runnable {
    val motor: DcMotor = opMode.hardwareMap.get(DcMotor::class.java, name)
    var RAMPING_COEFFICIENT = 0.8
    var MIN_SPEED = 0.2
        set(speed) {
            field = Math.abs(speed)
        }
    var MAX_SPEED = 1.0
        set(speed) {
            field = Math.abs(speed)
        }
    var MAX_NO_SLIP_SPEED = 1.0
        set(value) {
            field = Math.abs(value)
        }
    var target = 0.0
        set(target) {
            resetEncoder()
            this.current = currentPosition
            field = Math.abs(target)
        }
    var time = 0L
    private var current = 0.0
    var power
        set(value) {
            if(value==0.0 || !opMode.opModeIsActive()) {
                motor.power = 0.0
                return
            }
            when(rampingType) {
                RampingType.None -> {
                    this.current = currentPosition
                    motor.power = value
                    return
                }
                RampingType.ConstantJerk -> {
                    this.current = currentPosition
                    val k = 4.0 / target
                    val calcPower = k * current * (1 - (current/target)) * value + Double.MIN_VALUE
                    val expoPower = Range.clip(Math.pow(Math.abs(calcPower), RAMPING_COEFFICIENT), MIN_SPEED, MAX_SPEED)
                    if(value < 0) motor.power = -expoPower
                    if(value > 0) motor.power = expoPower
                    return
                }
                RampingType.Piecewise -> {
                    current = currentPosition
                    
                    if(target < RAMP_CLICKS_PROPORTION*CPR*2) { //no room to ramp
                        val calcSpeed = Math.min(Math.abs(value), MAX_NO_SLIP_SPEED)
                        motor.power = if(value>0) calcSpeed else -calcSpeed
                        return
                    }
                    if(current < RAMP_CLICKS_PROPORTION*CPR) { //Ramp up
                        val calcSpeed = Math.pow(current/CPR, 0.5)*(Math.abs(value)-MIN_SPEED)+MIN_SPEED
                        motor.power = if(value>0) calcSpeed else -calcSpeed
                        return
                    }
                    if(current > RAMP_CLICKS_PROPORTION*CPR && target - current > RAMP_CLICKS_PROPORTION*CPR) { //linear period
                        motor.power = value
                        return
                    }
                    if(target - current < RAMP_CLICKS_PROPORTION*CPR) { //ramp down
                        val calcSpeed = Math.pow((target-current)/CPR, 0.5)*(Math.abs(value)-0.325) + 0.325
                        motor.power = if(value>0) calcSpeed else -calcSpeed
                        return
                    }
                }
                RampingType.LinearDown -> {
                    current = currentPosition
                    val calcSpeed = (1 - current/target) * (Math.abs(value)-MIN_SPEED) + MIN_SPEED
                    motor.power = if(value>0) calcSpeed else -calcSpeed
                    return
                }
                RampingType.LinearUp -> {
                    current = currentPosition
                    val calcSpeed = (current/target) * (Math.abs(value)-MIN_SPEED) + MIN_SPEED
                    motor.power = if(value>0) calcSpeed else -calcSpeed
                    return
                }
            }
        }
        get() = motor.power
    var RAMP_CLICKS_PROPORTION = 0.5
    var targetPower = 0.0
    private var task = Tasks.Stop
    var runMode: DcMotor.RunMode
        set(runMode) {
            this.motor.mode = runMode                                                 
        }
        get() = this.motor.mode
    var model = MotorModel.NEVEREST40
    val CPR: Double
        get() = model.CPR
    val RPM: Double
        get() = model.RPM
    var currentPosition = 0.0
        private set
        get() = Math.abs(this.motor.currentPosition.toDouble())
    var zeroPowerBehavior: DcMotor.ZeroPowerBehavior
        set(zeroPowerBehavior) {
            this.motor.zeroPowerBehavior = zeroPowerBehavior
            motor.power = 0.0
        }
        get() = this.motor.zeroPowerBehavior
    var direction: DcMotorSimple.Direction
        set(value) {
            motor.direction = value
        }
        get() = motor.direction
        
    enum class MotorModel(val CPR: Double, val RPM: Double) {
        NEVEREST20(537.6, 333.3), 
        NEVEREST40(1120.0, 160.0), 
        NEVEREST60(1680.0, 106.6)
    }

    enum class Tasks {
        RunToPosition, RunForTime, Stop
    }

    enum class RampingType {
        None, ConstantJerk, Piecewise, LinearDown, LinearUp
    }
    var rampingType = RampingType.None
    
    init {
        this.model = MotorModel.NEVEREST40
        this.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        resetEncoder()
    }

    fun resetEncoder(): Motor {
        val initialBehavior = this.runMode
        this.runMode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        this.runMode = initialBehavior
        this.current = 0.0
        return this
    }

    fun runToPosition(targetPower: Double, targetClicks: Double, waitForCompletion: Boolean = true) {
        resetEncoder()
        isBusy = true
        this.target = Math.abs(targetClicks)
        this.targetPower = targetPower
        if (!waitForCompletion) {
            task = Tasks.RunToPosition
            val thread = Thread(this)
            thread.start()
        } else {
            while (!this.isAtTarget()) {
                if(!opMode.opModeIsActive()) {
                    return
                }
                this.power = targetPower
            }
            stopMotor()
        }
        isBusy = false
    }

    fun runForTime(targetPower: Double, time: Long, waitForCompletion: Boolean = true) {
        isBusy = true
        this.time = time
        this.targetPower = targetPower
        if(!waitForCompletion) {
            task = Tasks.RunForTime
            val thread = Thread(this)
            thread.start()
        } else {
            val runTime = ElapsedTime()
            motor.power = targetPower
            while(runTime.milliseconds() < this.time) if(!opMode.opModeIsActive()) continue
            this.stopMotor()
        }
        isBusy = false
    }

    fun stopMotor() {
        this.motor.power = 0.0
    }

    fun isAtTarget(): Boolean = Math.abs(currentPosition) + 30 >= Math.abs(target)

    override fun toString(): String =
        "Target clicks: $target\n" + 
        "Target power: $targetPower\n" +  
        "Current power: ${motor.power}\n"
        
    override fun run() {
        isBusy = true
        when(this.task) {
            Tasks.RunToPosition -> {
                while(!this.isAtTarget()) {
                    if(!opMode.opModeIsActive()) return                    
                    this.power = this.targetPower
                }
                this.stopMotor()
                task = Tasks.Stop
                isBusy = false
                Thread.currentThread().interrupt()
            }
            Tasks.RunForTime -> {
                val runTime = ElapsedTime()
                runTime.reset()
                this.motor.power = targetPower
                while(runTime.milliseconds()<this.time) if(!opMode.opModeIsActive()) continue
                this.stopMotor()
                task = Tasks.Stop
                isBusy = false
                Thread.currentThread().interrupt()
            }
            Tasks.Stop -> this.stopMotor()            
        }
    }
}